File: Data Structures & Algorithms/valid-tree/submission-5.py
class Solution:
    def validTree(self, n: int, edges: List[List[int]]) -> bool:
        if len(edges) != n - 1:
            return False
        graph = {i:[] for i in range(n)}
        for x, y in edges:
            graph[x].append(y)
            graph[y].append(x)
        visited, cycle = set(), set()
        def dfs(i, prev):
            if i in visited:
                return True
            if i in cycle:
                return False
            cycle.add(i)
            for n in graph[i]:
                if n == prev:
                    continue
                if not dfs(n, i):
                    return False
            cycle.remove(i)
            visited.add(i)
            return True
        for i in range(n):
            if not dfs(i, -1):
                return False
        return len(visited) == n
            

File: Data Structures & Algorithms/course-schedule/submission-5.cpp
class Solution {
public:
    vector<int> status;
    vector<vector<int>> graph;
    bool canFinish(int numCourses, vector<vector<int>>& prerequisites) {
        status.assign(numCourses, 0);
        graph.assign(numCourses, vector<int>());
        for (const auto& p : prerequisites) {
            graph[p[0]].push_back(p[1]);
        }
        // dfs
        for (int course = 0; course < numCourses; course++) {
            if (!dfs(course)) {
                return false;
            }
        }
        return true;
    }
    bool dfs(int course) {
        // 2: visted; it's good
        if (status[course] == 2) {
            return true;
        }
        // 1: being visited; circle found
        if (status[course] == 1) {
            return false;
        }
        status[course] = 1;
        for (const int nei : graph[course]) {
            if (!dfs(nei)) {
                return false;
            }
        }
        status[course] = 2;
        return true;
    }
};
